lightning/src/ln/peer_handler.rs

```diff
@@ -737,44 +755,54 @@ impl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, L: Deref, CMH: Deref> P
 
 	fn do_attempt_write_data(&self, descriptor: &mut Descriptor, peer: &mut Peer) {
 		while !peer.awaiting_write_event {
-			if peer.should_buffer_gossip_backfill() {
-				match peer.sync_status {
-					InitSyncTracker::NoSyncRequested => {},
-					InitSyncTracker::ChannelsSyncing(c) if c < 0xffff_ffff_ffff_ffff => {
-						if let Some((announce, update_a_option, update_b_option)) =
-							self.message_handler.route_handler.get_next_channel_announcement(c)
-						{
-							self.enqueue_message(peer, &announce);
-							if let Some(update_a) = update_a_option {
-								self.enqueue_message(peer, &update_a);
+			if peer.should_buffer_message_backfill() {
+				let next_onion_message_opt = if let Some(peer_node_id) = peer.their_node_id {
+					self.message_handler.onion_message_handler.next_onion_message_for_peer(peer_node_id)
+				} else { None };
+
+				// For now, let onion messages starve gossip.
+				if let Some(next_onion_message) = next_onion_message_opt {
+					self.enqueue_message(peer, &next_onion_message);
+				} else {
+					match peer.sync_status {
+						InitSyncTracker::NoSyncRequested => {},
+						InitSyncTracker::ChannelsSyncing(c) if c < 0xffff_ffff_ffff_ffff => {
+							if let Some((announce, update_a_option, update_b_option)) =
+								self.message_handler.route_handler.get_next_channel_announcement(c)
+							{
+								self.enqueue_message(peer, &announce);
+								if let Some(update_a) = update_a_option {
+									self.enqueue_message(peer, &update_a);
+								}
+								if let Some(update_b) = update_b_option {
+									self.enqueue_message(peer, &update_b);
+								}
+								peer.sync_status = InitSyncTracker::ChannelsSyncing(announce.contents.short_channel_id + 1);
+							} else {
+								peer.sync_status = InitSyncTracker::ChannelsSyncing(0xffff_ffff_ffff_ffff);
 							}
-							if let Some(update_b) = update_b_option {
-								self.enqueue_message(peer, &update_b);
+						},
+						InitSyncTracker::ChannelsSyncing(c) if c == 0xffff_ffff_ffff_ffff => {
+							if let Some(msg) = self.message_handler.route_handler.get_next_node_announcement(None) {
+								self.enqueue_message(peer, &msg);
+								peer.sync_status = InitSyncTracker::NodesSyncing(msg.contents.node_id);
+							} else {
+								peer.sync_status = InitSyncTracker::NoSyncRequested;
```

Comment:
I think since this, https://github.com/lightningdevkit/rust-lightning/pull/1604#discussion_r954358453 and https://github.com/lightningdevkit/rust-lightning/pull/1604#discussion_r954358655 are pre-existing I'd rather not address them in this PR to keep it small in scope. Maybe we should open an issue for them?

Reply:
yes definitely, if you provide your thoughts on those things, 
then i can go ahead and make those minor changes. :)

Reply:
Sadly github won't let me see what line this originally referred to. However, I think all the changes you suggested seem reasonable if you want to PR them, and we can discuss further on the PR :)